nexus/src/nexus.rs

```diff
@@ -1550,33 +1560,138 @@ impl Nexus {
             vpc.id(),
             subnet,
             mac,
-            ip,
-            params.clone(),
-        );
-        self.db_datastore.instance_create_network_interface(interface).await
+            params.identity.clone(),
+            params.ip,
+        )?;
+        let interface = self
+            .db_datastore
+            .instance_create_network_interface(interface)
+            .await
+            .map_err(NetworkInterfaceError::into_external)?;
+        Ok(interface)
+    }
+
+    /// Delete a network interface from the provided instance.
+    pub async fn instance_delete_network_interface(
+        &self,
+        opctx: &OpContext,
+        organization_name: &Name,
+        project_name: &Name,
+        instance_name: &Name,
+        interface_name: &Name,
+    ) -> DeleteResult {
+        let authz_project = self
+            .db_datastore
+            .project_lookup_by_path(organization_name, project_name)
+            .await?;
+        let (authz_instance, db_instance) = self
+            .db_datastore
+            .instance_fetch(opctx, &authz_project, instance_name)
+            .await?;
+        opctx.authorize(authz::Action::Modify, &authz_instance).await?;
+
+        // TODO-completeness: We'd like to relax this once hot-plug is supported
+        let stopped =
+            db::model::InstanceState::new(external::InstanceState::Stopped);
+        if db_instance.runtime_state.state != stopped {
+            return Err(external::Error::invalid_request(
+                "Instance must be stopped to detach a network interface",
+            ));
+        }
+        // TODO-cleanup: It's annoying that we need to look up the interface by
+        // name, which gets a single record, and then soft-delete that one
+        // record. We should be able to do both at once, but the
+        // `update_and_check` tools only operate on the primary key. That means
+        // we need to fetch the whole record first.
+        let interface = self
+            .db_datastore
+            .instance_lookup_network_interface(
+                &authz_instance.id(),
+                interface_name,
+            )
+            .await?;
+        self.db_datastore
+            .instance_delete_network_interface(&interface.id())
+            .await
+    }
+
+    /// Fetch a network interface attached to the given instance.
+    pub async fn instance_lookup_network_interface(
+        &self,
+        opctx: &OpContext,
+        organization_name: &Name,
+        project_name: &Name,
+        instance_name: &Name,
+        interface_name: &Name,
+    ) -> LookupResult<db::model::NetworkInterface> {
+        let authz_instance = self
+            .db_datastore
+            .instance_lookup_by_path(
+                organization_name,
+                project_name,
+                instance_name,
+            )
+            .await?;
+        opctx.authorize(authz::Action::ListChildren, &authz_instance).await?;
```

Comment:
I wasn't sure about this, so opted to have the auth checks be on the instance. I agree though that it could on the on the interface, and agree that I can add that in a separate auth-related PR for this stuff.

Reply:
Tracking this in #739